mujoco_py/mjpid.pyx

```diff
@@ -0,0 +1,97 @@
+from libc.math cimport fabs, fmax, fmin
+from mujoco_py.generated import const
+
+"""
+  Kp == Kp
+  Ki == Kp/Ti
+  Kd == Kp*Td
+
+  In this situation, Kp is a knob to tune the agressiveness, wheras Ti and Td will
+  change the response time of the system in a predictable way. Lower Ti or Td means
+  that the system will respond to error more quickly/agressively.
+
+  error deadband: ERROR_DEADBAND if set will shrink error within to 0.0
+  clamp on integral term: INTEGRAL_MAX_CLAMP helps on saturation problem in I.
+  derivative smoothing term: DERIVATIVE_GAIN_SMOOTHING reduces high frequency noise in D.
+
+  set in user="Ti Td iClamp errBand iSmooth" in mujoco xml.
+"""
+cdef enum USER_DEFINED_ACTUATOR_PARAMS:
+    PROPORTIONAL_GAIN = 0,
+    INTEGRAL_TIME_CONSTANT = 1,
+    INTEGRAL_MAX_CLAMP = 2,
+    DERIVATIVE_TIME_CONSTANT = 3,
+    DERIVATIVE_GAIN_SMOOTHING = 4,
+    ERROR_DEADBAND = 5,
+
+
+cdef enum USER_DEFINED_CONTROLLER_DATA:
+    INTEGRAL_ERROR = 0,
+    LAST_ERROR = 1,
+    DERIVATIVE_ERROR_LAST = 2,
+    NUM_USER_DATA_PER_ACT = 3,
+
+
+cdef mjtNum c_zero_gains(const mjModel* m, const mjData* d, int id) with gil:
+    return 0.0
+
+
+cdef mjtNum c_pid_bias(const mjModel* m, const mjData* d, int id) with gil:
+    cdef double dt_in_sec = m.opt.timestep
+    cdef double error = d.ctrl[id] - d.actuator_length[id]
+    cdef int NGAIN = int(const.NGAIN)
+
+    cdef double Kp = m.actuator_gainprm[id * NGAIN + PROPORTIONAL_GAIN]
+    cdef double error_deadband = m.actuator_gainprm[id * NGAIN + ERROR_DEADBAND]
+    cdef double integral_max_clamp = m.actuator_gainprm[id * NGAIN + INTEGRAL_MAX_CLAMP]
+    cdef double integral_time_const = m.actuator_gainprm[id * NGAIN + INTEGRAL_TIME_CONSTANT]
+    cdef double derivative_gain_smoothing = \
+        m.actuator_gainprm[id * NGAIN  + DERIVATIVE_GAIN_SMOOTHING]
+    cdef double derivate_time_const = m.actuator_gainprm[id * NGAIN + DERIVATIVE_TIME_CONSTANT]
```

Comment:
By DERIVATIVE_TIME_CONSTANT do you actually mean DERIVATIVE_TIME_CONSTANT_ID? Same comment for lines 44 - 50 and lines 20 - 25 and lines 29 - 32. What is NGAIN? Is it also an ID?

Reply:
DERIVATIVE_TIME_CONSTANT is the index for the field, NGAIN is the total number of parameters

Reply:
updated to IDX_